metricflow/model/validations/identifiers.py

```diff
@@ -119,3 +123,86 @@ def validate_model(model: UserConfiguredModel) -> List[ValidationIssue]:  # noqa
+
+    @staticmethod
+    def _get_sub_identifier_names(identifier: Identifier) -> Sequence[str]:
+        sub_identifier_names = []
+        sub_identifier: CompositeSubIdentifier
+        for sub_identifier in identifier.identifiers or []:
+            if sub_identifier.name:
+                sub_identifier_names.append(sub_identifier.name.element_name)
+            elif sub_identifier.ref:
+                sub_identifier_names.append(sub_identifier.ref)
+        return sub_identifier_names
+
+    @staticmethod
+    def _get_sub_identifier_context(data_source: DataSource) -> Sequence[SubIdentifierContext]:
+        contexts = []
+        for identifier in data_source.identifiers or []:
+            contexts.append(
+                SubIdentifierContext(
+                    data_source_name=data_source.name,
+                    identifier_reference=identifier.name,
+                    sub_identifier_names=tuple(IdentifierConsistencyRule._get_sub_identifier_names(identifier)),
+                )
+            )
+        return contexts
+
+    @staticmethod
+    @validate_safely(whats_being_done="running model validation to ensure identifiers have consistent sub-identifiers")
+    def validate_model(model: UserConfiguredModel) -> List[ValidationIssueType]:  # noqa: D
+        issues: List[ValidationIssueType] = []
+        # The expected sub identifiers that an identifier has.
+        identifier_to_sub_identifiers: OrderedDict[str, Tuple[str, ...]] = OrderedDict()
+        # The identifiers that have different sub-identifiers in different data sources.
+        identifiers_with_inconsistent_sub_identifiers = []
+
+        identifier_contexts = []
+        issues = []
+        for data_source in model.data_sources or []:
+            sub_identifier_contexts = IdentifierConsistencyRule._get_sub_identifier_context(data_source)
+            for sub_identifier_context in sub_identifier_contexts:
+                identifier_contexts.append(sub_identifier_context)
+
+                identifier_reference = sub_identifier_context.identifier_reference
+                if identifier_reference.element_name not in identifier_to_sub_identifiers:
+                    identifier_to_sub_identifiers[
+                        identifier_reference.element_name
+                    ] = sub_identifier_context.sub_identifier_names
+
+                elif (
+                    identifier_to_sub_identifiers[identifier_reference.element_name]
+                    != sub_identifier_context.sub_identifier_names
+                ):
+                    identifiers_with_inconsistent_sub_identifiers.append(identifier_reference)
+
+        for identifier_reference in identifiers_with_inconsistent_sub_identifiers:
+            for identifier_context in identifier_contexts:
+                if identifier_context.identifier_reference.element_name == identifier_reference.element_name:
+                    issues.append(
+                        ValidationWarning(
```

Comment:
`ValidationWarning` is non-blocking, right? Do we want to make this strict or is that a follow-up for later?

Reply:
@QMalcolm will be addressing the `ValidationWarning` behavior in a follow-up refactor, so let's leave this for now.